Source/Orts.Simulation/Simulation/Signalling/Signals.cs

```diff
@@ -1231,22 +1245,39 @@ private void AddWorldInfo()
             {
                 if (signal != null)
                 {
-                    foreach (SignalHead head in signal.SignalHeads)
+                    if (signal.isSignal)
                     {
+                        foreach (SignalHead head in signal.SignalHeads)
+                        {
 
-                        // get reference using TDB index from head
+                            // get reference using TDB index from head
 
-                        uint TDBRef = Convert.ToUInt32(head.TDBIndex);
-                        SignalRefObject thisRef;
+                            uint TDBRef = Convert.ToUInt32(head.TDBIndex);
+                            SignalRefObject thisRef;
 
-                        if (SignalRefList.TryGetValue(TDBRef, out thisRef))
+                            if (SignalRefList.TryGetValue(TDBRef, out thisRef))
+                            {
+                                uint signalIndex = thisRef.SignalWorldIndex;
+                                if (signal.WorldObject == null)
+                                {
+                                    signal.WorldObject = SignalWorldList[(int)signalIndex];
+                                }
+                                SignalRefList.Remove(TDBRef);
+                            }
+                        }
+                    }
+                    else
+                    {
+                        SignalHead head = signal.SignalHeads[0];
+                        int speedPostIndex;
+
+                        if (SpeedPostRefList.TryGetValue(head.TDBIndex, out speedPostIndex))
                         {
-                            uint signalIndex = thisRef.SignalWorldIndex;
-                            if (signal.WorldObject == null)
+                            if (signal.SpeedPostWorldObject == null)
                             {
-                                signal.WorldObject = SignalWorldList[(int)signalIndex];
+                                signal.SpeedPostWorldObject = SpeedPostWorldList[speedPostIndex];
```

Comment:
Indexing a list like this is a linear O(n) operation. A dictionary could accomplish the same thing in constant-time.

I realize you were copying the existing signal ref code, so I'm not going to press the issue. It might be something to revisit later.

Reply:
I am not really good at optimizing code. If you have any proposals, I will gladly take them into account.